mobilecoind/src/payments.rs

```diff
@@ -1082,6 +1165,201 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
+        fractional_outputs: &[(Amount, PublicAddress)],
+        min_fill_value: u64,
+        tombstone_block: BlockIndex,
+        fog_resolver_factory: &Arc<dyn Fn(&[FogUri]) -> Result<FPR, String> + Send + Sync>,
+        opt_memo_builder: Option<Box<dyn MemoBuilder + 'static + Send + Sync>>,
+        rng: &mut (impl RngCore + CryptoRng),
+        _logger: &Logger,
+    ) -> Result<SignedContingentInput, Error> {
+        // Check that we have at least one destination.
+        if required_outputs.is_empty() && fractional_outputs.is_empty() {
+            return Err(Error::TxBuild("Must have at least one destination".into()));
+        }
+
+        // Collect all required FogUris from public addresses, then pass to resolver
+        // factory
+        let fog_resolver = {
+            let change_address = from_account_key.subaddress(change_subaddress_index);
+            let fog_uris = core::slice::from_ref(&change_address)
+                .iter()
+                .chain(required_outputs.iter().map(|x| &x.1))
+                .chain(fractional_outputs.iter().map(|x| &x.1))
+                .filter_map(|x| extract_fog_uri(x).transpose())
+                .collect::<Result<Vec<_>, _>>()?;
+            fog_resolver_factory(&fog_uris).map_err(Error::Fog)?
+        };
+
+        let (mut ring, mut global_indices): (Vec<TxOut>, Vec<u64>) = ring.into_iter().unzip();
+
+        // Add the input to the ring.
+        let position_opt = ring.iter().position(|tx_out| *tx_out == utxo.tx_out);
+        let real_key_index = match position_opt {
+            Some(position) => {
+                // The input is already present in the ring.
+                // This could happen if ring elements are sampled randomly from the ledger.
+                position
+            }
+            None => {
+                // The input is not already in the ring.
+                if ring.is_empty() {
+                    // Append the input and its proof of membership.
+                    ring.push(utxo.tx_out.clone());
+                    global_indices.push(global_index);
+                } else {
+                    // Replace the first element of the ring.
+                    ring[0] = utxo.tx_out.clone();
+                    global_indices[0] = global_index;
+                }
+                // The real input is always the first element. This is safe because
+                // TransactionBuilder sorts each ring.
+                0
+            }
+        };
+
+        let membership_proofs: Vec<TxOutMembershipProof> = vec![Default::default(); ring.len()];
+
+        // Create input credentials
+        let input_credentials = {
+            assert_eq!(
+                ring.len(),
+                membership_proofs.len(),
+                "Each ring element must have a corresponding membership proof."
+            );
+
+            let public_key = RistrettoPublic::try_from(&utxo.tx_out.public_key).unwrap();
```

Comment:
Does this not work?
```suggestion
            let public_key = RistrettoPublic::try_from(&utxo.tx_out.public_key)?;
```

Reply:
it should work, I will fix the `build_tx_proposal` function also